students/yixingxu/lesson09/mailroom5.py

```diff
@@ -0,0 +1,162 @@
+import sys
+import os
+
+class Donor():
+    def __init__(self, name):
+        self._name = name
+        self._donations = []
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, donor_name):
+        self._name = donor_name
+
+    @property
+    def donations(self):
+        return self._donations
+
+    def add_donation(self, donor_donation):
+        self.donations.append(donor_donation)
+
+
+class Donors_Collection():
+    def __init__(self):
+        self._donors = []
+
+    @property
+    def donors(self):
+        return self._donors
+
+    def add_new_donor(self, donor):
+        self.donors.append(donor)
+
+    def get_donors_names(self):
+        name_list = []
+        for donor in self._donors:
+            name_list.append(donor.name)
+        return name_list
+
+
+    def add_donation_amount(self, name, donation_amount):
+        for donor in self._donors:
+            if donor.name == name:
+                donor.donations.append(donation_amount)
+                break
+
+    def create_summary(self):
+        summary_list = []
+        for donor in self._donors:
+            total = sum(donor.donations)
+            number_of_gifts = len(donor.donations)
+            average = total / number_of_gifts
+            summary_list.append([donor.name, '$', total, number_of_gifts, '$', average])
+        return summary_list
+
+    def create_a_report(self):
+        # create a summary list
+        summary_list = self.create_summary()
+        self.print_report(summary_list)
+
+    def print_report(self,a_list):
+        donor_name_width = 20
+        total_given_width = 15
+        num_gifts_width = 5
+        average_gift_width = 15
+        seperator_width = 2
+        title = ['Donor Name', '|', 'Total Given', '|', 'Num Gifts', '|', 'Average Gift']
+        divider = '-' * (
+                    donor_name_width + total_given_width + num_gifts_width + average_gift_width + seperator_width * 5)
+        print('{:<20}{:<2}{:<15}{:<2}{:<5}{:<2}{:<15}'.format(*title))
+        print(divider)
+        for item in a_list:
+            print('{:<20}{:2}{:15.2f}  {:8} {:>2}{:12.2f}\n'.format(*item))
+
+    def send_a_thank_you(self):
```

Comment:
It's better to separate the code that deals with user input from the code that handles the data so the input processing would work better inside a module level function